File: integrations/anthropic/python/src/purra_anthropic/_structured.py
"""Pure admission for the adapter's versioned native output dialect."""
from collections.abc import Mapping, Sequence
from purra.structured import StructuredOutputError

DIALECT = "purra.anthropic-json-schema/v1"
_ALLOWED = {"type", "properties", "required", "additionalProperties", "items", "enum", "description", "title"}

def validate(invocation):
    output = invocation.output_contract
    if output is None or output.mode == "local":
        return None
    def reject():
        raise StructuredOutputError("Native output dialect is unsupported", code="structured_output_mode_unsupported")
    if invocation.tools or invocation.request.protocol_capabilities.json_schema_level != "json_schema":
        reject()
    enum_count = 0
    property_count = 0
    def visit(node, depth):
        nonlocal enum_count, property_count
        if depth > 8 or set(node) - _ALLOWED:
            reject()
        kind = node.get("type")
        if not isinstance(kind, str):
            if not isinstance(kind, Sequence) or len(kind) != 2 or "null" not in kind:
                reject()
            kind = next(t for t in kind if t != "null")
        if kind == "object":
            props = node.get("properties", {})
            property_count += len(props)
            if property_count > 100 or node.get("additionalProperties") is not False or set(node.get("required", ())) != set(props):
                reject()
            for child in props.values():
                visit(child, depth + 1)
        elif "properties" in node or "required" in node or "additionalProperties" in node:
            reject()
        if kind == "array":
            if not isinstance(node.get("items"), Mapping):
                reject()
            visit(node["items"], depth + 1)
        elif "items" in node:
            reject()
        if "enum" in node:
            enum_count += len(node["enum"])
            if enum_count > 100 or any(isinstance(v, (Mapping, Sequence)) and not isinstance(v, str) for v in node["enum"]):
                reject()
    visit(output.schema, 0)
    return DIALECT

File: integrations/anthropic/typescript/src/structured.ts
import { StructuredOutputError } from "purra";
import type { ModelCapabilitySnapshot, ModelRequest, JsonValue } from "purra";
const allowed = new Set(["type", "properties", "required", "additionalProperties", "items", "enum", "description", "title"]);
export function validateOutput(request: ModelRequest, capabilities: ModelCapabilitySnapshot): string | undefined {
  const output = request.outputContract;
  if (output === undefined || output.mode === "local") return undefined;
  function reject(): never { throw new StructuredOutputError("structured_output_mode_unsupported", "nativeDialect"); }
  if (request.tools.length || capabilities.protocol.jsonSchemaLevel !== "json_schema"
    || request.capabilitySnapshot?.protocol.jsonSchemaLevel !== "json_schema") reject();
  let enumCount = 0, propertyCount = 0;
  function visit(node: Readonly<Record<string, JsonValue>>, depth: number): void {
    if (depth > 8 || Object.keys(node).some(key => !allowed.has(key))) reject();
    let kind = node.type;
    if (typeof kind !== "string") {
      if (!Array.isArray(kind) || kind.length !== 2 || !kind.includes("null")) reject();
      kind = kind.find(t => t !== "null")!;
    }
    if (kind === "object") {
      const props = (node.properties ?? {}) as Readonly<Record<string, Readonly<Record<string, JsonValue>>>>;
      propertyCount += Object.keys(props).length;
      const required = (node.required ?? []) as readonly JsonValue[];
      if (propertyCount > 100 || node.additionalProperties !== false || !Array.isArray(required)
        || required.length !== Object.keys(props).length
        || Object.keys(props).some(key => !required.includes(key))) reject();
      for (const child of Object.values(props)) visit(child, depth + 1);
    } else if (["properties", "required", "additionalProperties"].some(key => key in node)) reject();
    if (kind === "array") {
      if (!node.items || typeof node.items !== "object" || Array.isArray(node.items)) reject();
      visit(node.items as Readonly<Record<string, JsonValue>>, depth + 1);
    } else if ("items" in node) reject();
    if (Array.isArray(node.enum)) {
      enumCount += node.enum.length;
      if (enumCount > 100 || node.enum.some(v => v !== null && typeof v === "object")) reject();
    }
  }
  visit(output.schema, 0);
  return "purra.anthropic-json-schema/v1";
}
